### ISSUE TITLE
dnf cache is not shared among builds

### ISSUE THREAD

**Issue by ampresent** [OSInside/kiwi#2491]
## Problem description
For DNF-based target systems, run `kiwi system build` multiple times.
DNF packages are re-downloaded on every build, and duplicate packages accumulate in /var/cache/kiwi.

## Expected behaviour
- DNF packages needn't to be re-downloaded
- DNF packages only keeps one copy in /var/cache/kiwi

## Steps to reproduce the behaviour
For DNF-based target systems, run `kiwi system build` multiple times.

## OS and Software information

* KIWI version: 9.25.21
* Operating system host version: CentOS 7.9
* Operating system target version: CentOS 8.5
* Open Build Service version (N/A if not using OBS): N/A
* Koji version (N/A if not using Koji): N/A

**Comment by ampresent** [comment]
This issue also occurs on the host system RockyLinux 9.2

**Comment by schaefi** [comment]
Hmm, that sounds not good. From a log of an integration test, this is the way we call dnf

1. bootstrap phase

```
EXEC: [dnf --config /usr/src/packages/KIWI-oem/build/image-root/kiwi_mgjg4uc1 -y --releasever=0 makecache]
[   25s] [ DEBUG   ]: 23:31:33 | EXEC: [dnf --config /usr/src/packages/KIWI-oem/build/image-root/kiwi_mgjg4uc1 -y --installroot /usr/src/packages/KIWI-oem/build/image-root --releasever=0 --setopt=install_weak_deps=False install basesystem dnf fedora-release filesystem grub2-efi-x64 grub2-efi-x64-modules shim]
```

2. install phase (chroot)

```
EXEC: [chroot /usr/src/packages/KIWI-oem/build/image-root dnf --config /kiwi_mgjg4uc1 -y --releasever=0 --setopt=install_weak_deps=False install NetworkManager btrfs-progs dhclient dracut-kiwi-oem-dump dracut-kiwi-oem-repart glibc-all-langpacks grub2 grub2-breeze-theme grubby kernel plymouth-theme-charge selinux-policy-targeted tzdata vim]
```

@Conan-Kudo out of your mind, do you see any issue ?

**Comment by schaefi** [comment]
> duplicate packages accumulate in /var/cache/kiwi

Can you share the contents or listing of /var/cache/kiwi ? Thanks

**Comment by Conan-Kudo** [comment]
I've never seen this before on my own system, so I'm curious what's going on here...

**Comment by ampresent** [comment]
Output of command `sudo du -hs  /var/cache/kiwi/dnf/cache/* | sort -h`  is like
```
....
3.1M    /var/cache/kiwi/dnf/cache/4160c418901d42659bdf7ff65283028c-677dee7b93891601
3.1M    /var/cache/kiwi/dnf/cache/b78d14ed780b4bd0b6db32a28a533b1d-677dee7b93891601
3.5M    /var/cache/kiwi/dnf/cache/054949844b69406b969add3bc063cade-7900815317e07afc
3.5M    /var/cache/kiwi/dnf/cache/0d46357f7b7a4d2cae82baf8f7fd8dd3-7900815317e07afc
4.0M    /var/cache/kiwi/dnf/cache/149bbb414ea0430f8c8a321288e9313b.solv
4.0M    /var/cache/kiwi/dnf/cache/8e2780955a274f62b3a7e39df8130393.solv
6.3M    /var/cache/kiwi/dnf/cache/26ec18149fc54c6caec3b8ba232c123f-filenames.solvx
6.3M    /var/cache/kiwi/dnf/cache/c9158a893b98408a8c074ce6d98122d1-filenames.solvx
9.8M    /var/cache/kiwi/dnf/cache/149bbb414ea0430f8c8a321288e9313b-updateinfo.solvx
9.8M    /var/cache/kiwi/dnf/cache/8e2780955a274f62b3a7e39df8130393-updateinfo.solvx
13M     /var/cache/kiwi/dnf/cache/149bbb414ea0430f8c8a321288e9313b-filenames.solvx
13M     /var/cache/kiwi/dnf/cache/8e2780955a274f62b3a7e39df8130393-filenames.solvx
18M     /var/cache/kiwi/dnf/cache/149bbb414ea0430f8c8a321288e9313b-4f8fc4485131eb3a
18M     /var/cache/kiwi/dnf/cache/8e2780955a274f62b3a7e39df8130393-4f8fc4485131eb3a
114M    /var/cache/kiwi/dnf/cache/26ec18149fc54c6caec3b8ba232c123f-0927ee5e32f3f19a
114M    /var/cache/kiwi/dnf/cache/c9158a893b98408a8c074ce6d98122d1-0927ee5e32f3f19a
469M    /var/cache/kiwi/dnf/cache/4e67d343f454417c8e91f139fdfd9e0f-8bae5e100d465154
469M    /var/cache/kiwi/dnf/cache/7f74006e4fa641599746dc1c19654a66-8bae5e100d465154
```

It seems there is a pattern where dnf stores caches, which goes like `/var/cache/kiwi/dnf/cache/<build-id>-<repo id>`

**Comment by schaefi** [comment]
Hmm, and this ID's are created new with each build ? if so it's hard to re-use data from former caches

**Comment by Conan-Kudo** [comment]
@kontura can verify this, but I believe this is actually versioning to ensure libsolv solver cache invalidation works correctly. It should be stable with the same version of DNF being used each time.

**Comment by kontura** [comment]
I don't think it has anything to do with libsolv cache invalidation, the checksums for that are stored inside the `*.solv` files.

For example dnf cache for `fedora` repository looks like:
```
/var/cache/dnf/fedora-6c3a9e5977a00788
/var/cache/dnf/fedora-6c3a9e5977a00788/repodata
/var/cache/dnf/fedora-6c3a9e5977a00788/repodata/repomd.xml
/var/cache/dnf/fedora-6c3a9e5977a00788/repodata/b5d49b491109ae304a7639c9a0e66d17b3d6543a30f02f3e289b21435f88d0f5-primary.xml.zck
/var/cache/dnf/fedora-6c3a9e5977a00788/repodata/89e378a0348f74ae7a35b641eddafee238c4baa3d157a9ec3a74659d9c2e81f3-filelists.xml.zck
/var/cache/dnf/fedora-6c3a9e5977a00788/repodata/ea16c5e212b0c6a209e858034ce8c6c3bcf43c669e45a78000fa709bb5521309-comps-Everything.x86_64.xml.gz
/var/cache/dnf/fedora-6c3a9e5977a00788/metalink.xml
/var/cache/dnf/fedora-filenames.solvx
/var/cache/dnf/fedora.solv
```

That means that in your case the longer hex strings are repository ids.
You have a repo with id `149bbb414ea0430f8c8a321288e9313b`, `8e2780955a274f62b3a7e39df8130393`,...

**Comment by schaefi** [comment]
Hmm, I'm sorry I cannot reproduce this issue. Here is what I did. I built the fedora example from the community git here:

* https://github.com/OSInside/kiwi-descriptions/tree/main/fedora/x86_64/fedora-37.0

with the following kiwi call

```
rm -rf /var/cache/kiwi/dnf
sudo kiwi-ng system build --description /home/ms/Project/kiwi-descriptions/fedora/x86_64/fedora-37.0/ --target-dir /tmp/mytest
```
This was the initial call and I stored the build log. It has entries of the form

```
[ DEBUG   ]: 09:26:19 | bootstrap: Install  184 Packages
[ DEBUG   ]: 09:26:19 | bootstrap: Total download size: 101 M
[ DEBUG   ]: 09:26:19 | bootstrap: Installed size: 513 M
[ DEBUG   ]: 09:26:19 | bootstrap: Downloading Packages:
[ DEBUG   ]: 09:26:19 | bootstrap: (1/184): ...
...
[ DEBUG   ]: 09:27:19 | system: Install  167 Packages
[ DEBUG   ]: 09:27:19 | system: Total download size: 172 M
[ DEBUG   ]: 09:27:19 | system: Installed size: 393 M
[ DEBUG   ]: 09:27:19 | system: Downloading Packages:
[ DEBUG   ]: 09:27:19 | system: (1/167): ...
```

so far so good. Then I repeat the build and kept the cache

```
rm -rf /tmp/mytest
sudo kiwi-ng system build --description /home/ms/Project/kiwi-descriptions/fedora/x86_64/fedora-37.0/ --target-dir /tmp/mytest
```

In the build log I see that it takes the cache information

```
[ DEBUG   ]: 09:31:34 | bootstrap: Install  184 Packages
[ DEBUG   ]: 09:31:34 | bootstrap: Total size: 101 M
[ DEBUG   ]: 09:31:34 | bootstrap: Installed size: 513 M
[ DEBUG   ]: 09:31:34 | bootstrap: Downloading Packages:
[ DEBUG   ]: 09:31:34 | bootstrap: [SKIPPED] basesystem-11-14.fc37.noarch.rpm: Already downloaded                 
[ DEBUG   ]: 09:31:34 | bootstrap: [SKIPPED] ... and so on
...
[ DEBUG   ]: 09:32:11 | system: Install  167 Packages
[ DEBUG   ]: 09:32:11 | system: Total size: 172 M
[ DEBUG   ]: 09:32:11 | system: Installed size: 393 M
[ DEBUG   ]: 09:32:11 | system: Downloading Packages:
[ DEBUG   ]: 09:32:11 | system: [SKIPPED] augeas-libs-1.13.0-3.fc37.x86_64.rpm: Already downloaded
[ DEBUG   ]: 09:32:11 | system: [SKIPPED] ... and so on
```

So it build completely from cache. The cache looks like this

```
/var/cache/kiwi/dnf/
├── cache
│   ├── expired_repos.json
│   ├── Fedora37-a3256cee7c7d69ad
│   │   ├── metalink.xml
│   │   ├── packages
│   │   │   ├── augeas-libs-1.13.0-3.fc37.x86_64.rpm
│   │   │   ├── basesystem-11-14.fc37.noarch.rpm
│   │   │   ├── boost-system-1.78.0-9.fc37.x86_64.rpm
│   │   │   ├── boost-thread-1.78.0-9.fc37.x86_64.rpm
│   │   │   ├── zypp-plugins-17.31.1-1.fc37.noarch.rpm
│   │   │   └── ... and many more
│   │   └── repodata
│   │       ├── 721f55ca56291adc1d1a566c06716b12c2373e57b213c6da778a9349e17e312a-comps-Everything.x86_64.xml
│   │       ├── 72c5f1d9a2866f4975c46f79ddc144e41a96992f842b8522e97603c0650a4d03-primary.xml.zck
│   │       ├── 817f980753a2031de30d066e009ffda338bad185a7dc7c617e5e6c33f869455c-filelists.xml.zck
│   │       ├── gen
│   │       │   └── groups.xml
│   │       └── repomd.xml
│   ├── Fedora37-filenames.solvx
│   ├── Fedora37.solv
│   ├── Fedora37-updates-fd4d3d0d1c34d49a
│   │   ├── metalink.xml
│   │   ├── packages
│   │   │   ├── alternatives-1.24-1.fc37.x86_64.rpm
│   │   │   ├── audit-libs-3.1.2-5.fc37.x86_64.rpm
│   │   │   ├── authselect-1.4.2-1.fc37.x86_64.rpm
│   │   │   └── ... and many more
│   │   └── repodata
│   │       ├── 3a72af5d1120b6adc68c06e2c4eeafdfeaa23cb8dd7765fba533b5f7ac7ae76b-comps-Everything.x86_64.xml.gz
│   │       ├── 687bb473a8403e32ef7e023f0b32b894b6e6297ef649e76afc00ee96d032093b-prestodelta.xml.zck
│   │       ├── 7580dffaa133fad14bb8e8f27f959a16855ffb77f334cba639aac553d9763490-filelists.xml.zck
│   │       ├── 961cf374cee4102b74f413bea2259411ef8a9c6d8ac69e76803d5857472894ce-primary.xml.zck
│   │       ├── 96a90c3b769f5f60e375063bb855ceadafe7a75390511ce83d8e517b6defe18b-updateinfo.xml.zck
│   │       ├── gen
│   │       │   └── groups.xml
│   │       └── repomd.xml
│   ├── Fedora37-updates-filenames.solvx
│   ├── Fedora37-updates-presto.solvx
│   ├── Fedora37-updates.solv
│   └── Fedora37-updates-updateinfo.solvx
├── pluginconf
│   ├── priorities.conf
│   └── versionlock.conf
├── repos
│   ├── Fedora37.repo
│   └── Fedora37-updates.repo
└── vars
```

So from my perspective everything worked as expected.

dnf on the host (bootstrap):

```
dnf --version
4.10.0
  Installed: dnf-0:4.10.0-bp155.3.10.noarch at Mon 15 Jan 2024 08:16:34 AM GMT
  Built    : https://bugs.opensuse.org at Mon 22 May 2023 10:37:22 AM GMT

  Installed: rpm-0:4.14.3-150400.59.3.1.x86_64 at Wed 15 Nov 2023 02:27:07 PM GMT
  Built    : https://www.suse.com/ at Mon 14 Aug 2023 10:40:37 AM GMT
```

dnf in the image (system):

```
sudo chroot /tmp/mytest/build/image-root/ dnf --version
4.18.0
  Installed: dnf-0:4.18.0-2.fc37.noarch at Fri 01 Mar 2024 08:31:38 AM GMT
  Built    : Fedora Project at Thu 19 Oct 2023 09:04:15 AM GMT

  Installed: rpm-0:4.18.2-1.fc37.x86_64 at Fri 01 Mar 2024 08:31:37 AM GMT
  Built    : Fedora Project at Mon 13 Nov 2023 03:50:44 PM GMT
```

**Comment by schaefi** [comment]
One part I could imagine to be the issue is the repo definition. I used a repo definition with alias like this:

```xml
<repository type="rpm-md" alias="Fedora37" sourcetype="metalink">
    <source path="https://mirrors.fedoraproject.org/metalink?repo=fedora-37&amp;arch=x86_64"/>
</repository>
```

The alias name is used for the repo file. So in the cache you see ```Fedora37``` and that is a static information.

If you use a repo definition without an alias kiwi calculates a hash over the provided source path. If the source path is a metalink or any other form of dynamic information that calculated id can vary between builds and that will cause a new repo file and a new cache for it.

So maybe your issue is solved if you just give your repos an alias name ?

**Comment by raphaeltraviss** [comment]
I had a similar issue with openSUSE/zypper, and the alias solved it in my case (version `10.0.8`)

Non-working config--Zypper was downloading every single package every time I ran `system prepare`:
```
  <repository
    imageinclude="true"
  >
    <source path="http://download.opensuse.org/distribution/leap/15.5/repo/oss/" />
  </repository>

  <repository
    imageinclude="true"
  >
    <source path="http://download.opensuse.org/update/leap/15.5/oss/" />
  </repository>

  <repository
    imageinclude="true"
  >
    <source path="http://download.opensuse.org/distribution/leap/15.5/repo/non-oss/" />
  </repository>

  <repository
    imageinclude="true"
  >
    <source path="http://download.opensuse.org/update/leap/15.5/non-oss/" />
  </repository>
```

Working XML definition with the alias--packages are cached on the host and reused on each `system prepare` of a new image directory:
```
  <repository
    imageinclude="true"
    alias="one"
  >
    <source path="http://download.opensuse.org/distribution/leap/15.5/repo/oss/" />
  </repository>

  <repository
    imageinclude="true"
    alias="two"
  >
    <source path="http://download.opensuse.org/update/leap/15.5/oss/" />
  </repository>

  <repository
    imageinclude="true"
    alias="three"
  >
    <source path="http://download.opensuse.org/distribution/leap/15.5/repo/non-oss/" />
  </repository>

  <repository
    imageinclude="true"
    alias="four"
  >
    <source path="http://download.opensuse.org/update/leap/15.5/non-oss/" />
  </repository>
```

The repo directory names were indeed just hashes, filling up the cache directory, and never being reused:
![image](https://github.com/OSInside/kiwi/assets/2137944/14e90678-aaf4-4a9b-b6ea-a2bb72a8ba86)

**Comment by schaefi** [comment]
> I had a similar issue with openSUSE/zypper, and the alias solved it in my case (version 10.0.8)

great, thanks for the feedback.
